DE-6746: Add a GH workflow to publish a release

fingerartur opened the pull request:
Here I'm adding a GitHub workflow to publish a new version to NPM and deploy API docs for that version to https://players.castlabs.com/react-dom/x.x.x/docs/. It gets triggered when we create a new Release in GitHub (when creating the release, select "Create new tag" and specify the target version).

Beta versions are also supported and they will be published accordingly to a NPM beta channel (non-main channel).

thasso commented on .github/workflows/publish.yml:
```diff
@@ -0,0 +1,76 @@
+
+      - name: Write NPM RC File
+        run: |
+          echo '@castlabs:registry=https://registry.npmjs.org' > .npmrc
```
This one gives me unease :) Can we 1) Use the secret directly here and 2) add a finally step that always executes and removes the .npmrc file?

fingerartur replied:
Yeah, I'll pass it directly to the shell that needs it.

fingerartur replied:
Actually ... I've been bamboozled. I think we both may have misunderstood what this step does - it does **_not_** write out the NPM_TOKEN to that file, it creates a templated .npmrc file, and NPM then fills that template with the current value of `$NPM_TOKEN` at runtime, meaning the token is in RAM all that time, it does not get written to a file. ... So I think we are happy with this step security-wise.

So I will keep the logic of this step, but I will move it to the `.github/actions/publish/main.js` action, which does the publishing to NPM.

thasso commented on story/stories/examples/DefaultSkinSource.tsx:
```diff
@@ -14,7 +15,7 @@ import '@castlabs/prestoplay-react-components/dist/themes/pp-ui-base-theme-embed
 const baseConfig = {
   // Replace this by your license key
-  license: '7c8165eb7231508ce4eae02181d53b1c',
+  license: '7c816...81d53b1c',
```
Maybe we add a comment here that leaving this out will use the embedded dev license or even make it empty and just have the comment so it "just works" for local development?

thasso commented on story/stories/examples/YoutubeSkinSource.ts:
```diff
@@ -32,16 +24,14 @@ import 'youtube-style.css'
+
+const baseConfig = {
+  // Replace this by your license key
+  license: '7c8165...81d53b1c',
```
s.o.

fingerartur replied:
Fair warning: My German language skills suck incredibly 😄

fingerartur commented on .github/actions/publish/main.js:
```diff
@@ -0,0 +1,33 @@
+  const betaLabel = betaMatch[1];
+  core.info(`Publishing a Beta version ${version}.`)
+  execSync(`npm publish ${dryRun ? '--dry-run' : ''} --tag ${betaLabel}`, { stdio: 'inherit' })
+  process.exit(0);
```
Ok I'll get rid of those process.exit(0);

fingerartur commented:
@thasso I've also added provenance verification and switched to a granular NPM token.

Provenance works, I've tested it with a beta NPM release, just an interesting note on that topic: it needs Node.js version 20 (actually version 18 would be enough, but GitHub actions somehow don't support it, at least not explicitly, so version 20 it is 😄).

thasso commented on .github/actions/publish/main.js:
```diff
@@ -1,33 +1,72 @@
+
+try {
+  fs.writeFileSync(utils.file('.npmrc'), npmRcContent)
+  execSync(`export NPM_TOKEN=${npmToken}; npm publish ${args.join(' ')}`, { stdio: 'inherit' })
```
I think you do not even need to `export` the token here. I think you can either _pass_ this for that one shell commend, i.e. `NPM_TOKEN="" npm publish...` or (and I tend to like that more) pass it as an `env` option in the `options` of `execSync`.

fingerartur replied:
Nice, I will use `env`.

thasso reviewed (changes requested):
nice, just one more suggestion on how we expose the token that you might want to give a shot.
